File: 581-shortest-unsorted-continuous-subarray/581-shortest-unsorted-continuous-subarray.cpp
class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
        const int n=nums.size();
        stack<int>st;
       int first=n;
        
        for(int i=0;i<n;i++){
            while(!st.empty() && nums[st.top()] > nums[i]){
                first=min(first, st.top());
                st.pop();
            }
          st.push(i);           
        }
        
        
         while(!st.empty()){
             st.pop();
         }
        
        int lst=-1;
           for(int i=n-1;i>=0;--i){
            while(!st.empty() && nums[st.top()] < nums[i]){
               lst=max(lst, st.top());
                st.pop();
            }
        st.push(i);
        }

    return lst>first?lst-first+1:0;
    
    }
};



    

File: 581-shortest-unsorted-continuous-subarray/581-shortest-unsorted-continuous-subarray.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        
        stack = []
        first = len(nums)
        for i in range(len(nums)):
            while stack != [] and nums[stack[-1]] > nums[i]:
                first = min(first, stack.pop())
               
              
            stack.append(i)    
            
            
        stack.clear()
        
        last=-1
        for i in range(len(nums)-1,-1,-1):
            while (stack and nums[stack[-1]] < nums[i]):
                last=max(last, stack.pop())
                
            stack.append(i)
            
       
        if last < first:
            return 0
        else:
            return last-first+1  